src/main/java/seedu/address/model/note/TagNoteMap.java

```diff
@@ -0,0 +1,116 @@
+package seedu.address.model.note;
+
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import seedu.address.model.client.Client;
+import seedu.address.model.tag.Tag;
+
+
+/**
+ * Manages the relationship between Tags and Notes.
+ */
+
+public class TagNoteMap {
+    //    TODO: Add implementation
+    public final Set<Tag> tagList = new HashSet<>(); // probably redundant todo: remove later
+    public final Set<Note> noteSet = new HashSet<>();
+    public final Map<Tag, Set<Note>> tagToNotesMap = new HashMap<>();
+    public final Map<Note, Set<Tag>> noteToTagsMap = new HashMap<>();
+
+
+    /**
+     * Initialises the TagNoteMap with a list of clients.
+     *
+     * @param clients The list of clients containing ClientNotes and associated tags.
+     */
+    public void initMapFromClients(List<Client> clients) {
+        for (Client client : clients) {
+            Set<Note> clientNotes = client.getClientNotes();
+            noteSet.addAll(clientNotes);
+            for (Note clientNote : clientNotes) {
+                Set<Tag> tags = clientNote.getTags();
+                tagList.addAll(tags);
+                updateTagsForNote(tags, clientNote);
+            }
+        }
+        // todo: how to remove old entries to the map?
+    }
+
+    /**
+     * Initialises the TagNoteMap with a list of countries.
+     *
+     * @param countryNotes The set of countries containing ClientNotes and associated tags.
+     */
+    public void initMapFromCountries(Set<Note> countryNotes) {
+        // todo: make init work when passed in a list of countries
+        noteSet.addAll(countryNotes);
+        for (Note countryNote : countryNotes) {
+            Set<Tag> tags = countryNote.getTags();
+            tagList.addAll(tags);
+            updateTagsForNote(tags, countryNote);
+        }
+    }
+
+    // todo: how to remove old entries to the map? <== need to do this to "sync"
+    //       with the current model right?
+    public Set<Tag> getTagsForNote(Note note) {
+        return Collections.unmodifiableSet(noteToTagsMap.getOrDefault(note, new HashSet<>()));
+    }
+
+    public Set<Note> getNotesForTag(Tag tag) {
+        return Collections.unmodifiableSet(tagToNotesMap.getOrDefault(tag, new HashSet<>()));
+    }
+
+    /**
+     * Links a new set of tags to a note.
+     *
+     * @param newTags The tags to associate with a particular note.
+     * @param note    The note to associate the tag with.
+     */
+    public void updateTagsForNote(Set<Tag> newTags, Note note) {
```

Comment:
should probably requireAllNonNull here.
Also, is there a need for this method to be public? It seems to be only used here

Reply:
added reason for the `public` access but I've also added a todo because I expect changes to this after storage is done.